### PR TITLE
Add KubeCon NA 2019 blog post

### PR DESCRIPTION
**Pull request by gabrieljackson**
A blog post with highlights from KubeCon NA 2019.

https://mattermost.atlassian.net/browse/MM-20799

### CODE REVIEW

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+It goes without saying that we are just trying to follow an increasingly-common trend here. For this reason, I was happily surprised with the concepts presented by Goutham Veeramachaneni & Edward Welch in this talk:
+
+{{< youtube BmPnNmN9jtc >}}
+<br/>
+
+Their idea of a Monomicrolith really resonated. One of the struggles we have with the cloud server is that its responsibility to maintain our various cloud objects require it to be the source of truth in when creating, updating, or tearing down infrastructure. This, in turn, means that a lot of configuration is defined through the server. The talk explored the idea that this sort of design isn’t necessarily a flaw. Instead, if the code is modular enough, you can specify what pieces of functionality each process should run on startup. This allows for most of the benefits of a single code-base and configuration style, while still allowing for extreme flexibility and horizontal scaling.
```
Comment (jwilander) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
Their idea of a Monomicrolith really resonated. One of the struggles we have with the cloud server is that its responsibility to maintain our various cloud objects require it to be the source of truth when creating, updating, or tearing down infrastructure. This, in turn, means that a lot of configuration is defined through the server. The talk explored the idea that this sort of design isn’t necessarily a flaw. Instead, if the code is modular enough, you can specify what pieces of functionality each process should run on startup. This allows for most of the benefits of a single code-base and configuration style, while still allowing for extreme flexibility and horizontal scaling.
```

Above changes `require it to be the source of truth in when creating, updating,` to `require it to be the source of truth when creating, updating,`

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+slug: kubecon-na-2019
+date: 2019-12-18T12:00:00-04:00
+categories:
+    - "kubernetes"
+author: Gabe Jackson
+github: gabrieljackson
+community: gabe.jackson
+---
+
+It wasn’t my first time flying. It also wasn’t the first time I had arrived at the airport early enough to require multiple cups of coffee. As I sat at my gate, however, I did find myself wondering for the first time: had I ever heard the words “lambda,” “stateful set,” and “yaml” while waiting to board?
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
It wasn’t my first time flying. It also wasn’t the first time I had arrived at the airport early enough to require multiple cups of coffee. As I sat at my gate, however, I did find myself wondering for the first time: had I ever heard the words “Lambda,” “stateful set", and “yaml” while waiting to board?
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+You see, a bunch of us were coming from all over to attend KubeCon + CloudNativeCon North America 2019. Now that the conference has come and gone, let’s dive into some of our takeaways.
+
+#### A container-sized intro
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
#### A Container-sized Intro
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+Every year since the [first Kubernetes GitHub commit](https://github.com/kubernetes/kubernetes/commit/2c4b3a562ce34cddc3f8218a2c4d11c7310e6d56) in 2014, the project has needed less of an introduction. At its core, Kubernetes is an open source platform for running containerized workloads and services. With thousands of unique contributors and the support of some very large organizations, it has matured to the point where it runs varied complex workloads all over the world. This is, partly, due to the fact that Kubernetes is extensible in many ways and can be tailored to provide rich interactions and support for specific use-cases. We at Mattermost also find this incredibly valuable.
+
+With all that in mind, KubeCon is a conference where the community around Kubernetes comes together to talk about what is next, share lessons learned, compare experiences using the platform, and to simply connect.
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
With all that in mind, KubeCon is a conference where the community around Kubernetes comes together to share lessons learned, compare experiences using the platform, talk about what's next, and simply connect.
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+#### Kubernetes is not slowing down
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
#### Kubernetes is Not Slowing Down
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+One of the most difficult aspects of attending Kubecon can simply be deciding which talks to attend. Nearly every time slot had a diverse set of topics which demonstrates that the platform is still being enhanced in nearly every technical direction. Everything from storage, networking, security, performance, monitoring, observability, and extensibility continues to change every year to meet new demands. There were also multiple amazing talks about how _not_ to do things, and why -- which are sometimes more useful than hearing what you _should_ be doing.
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
One of the most difficult aspects of attending KubeCon is deciding which talks to attend. Nearly every time-slot had a diverse set of topics which demonstrates that the platform is still being enhanced in nearly every technical direction. 

Everything from storage, networking, security, performance, monitoring, observability, and extensibility continues to change every year to meet new demands. There were also multiple amazing talks about how _not_ to do things, and why -- which are sometimes more useful than hearing what you _should_ be doing.
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+{{< youtube FrQ8Lwm9_j8 >}}
+<br/>
+
+Additionally, version 1.16 of Kubernetes was recently released. Despite the platform regularly receiving quarterly minor version upgrades, a lot has been packed into the latest edition.
+
+In particular, we were excited that custom resources reached general availability in 1.16. Although this functionally doesn’t mean much, it shows that extensibility really is an important aspect of the platform. To break it down a bit, a custom resource definition allows the users of Kubernetes platforms to define their own specifications for resources. These specifications are then treated as normal Kubernetes API objects. This allows for something like a database, monitoring service, backend framework, etc. to be easily provisioned in a cluster based on these definitions. We use this to define Mattermost installations for our [operator to work with](https://github.com/mattermost/mattermost-operator).
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
In particular, we were excited that custom resources reached general availability in 1.16. 

A custom resource definition allows the users of Kubernetes platforms to define their own specifications for resources. These specifications are then treated as normal Kubernetes API objects. This allows for something like a database, monitoring service, backend framework, etc. to be easily provisioned in a cluster based on these definitions. We use this to define Mattermost installations for our [operator to work with](https://github.com/mattermost/mattermost-operator). Although functionally this doesn’t mean much, it shows that extensibility really is an important aspect of the platform. 
```

Comment (justinegeffen) [reply site/content/blog/2019-12-18-kubecon-na-2019.md]: It sounds like custom resource definition is functionally useful. I'm wondering if there's a better way to describe it? Maybe something like "although this doesn't impact the day-to-day functionality..."?

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+By leveraging custom resources, Kubernetes principles such as the control loop, and custom code, operators aim to provide the experience of a skilled maintainer of the service or software for which they are responsible. In essence, this allows not only the possibility of easy deployment of a given product or service, but also continued software health monitoring, upgrades, advanced deployment schemes, alerting, metrics, scaling, and more. Operators run alongside resources they manage in the same Kubernetes clusters, giving them direct control and monitoring of said resources. Think of them as being coded with the knowledge of a system admin that can make changes and manage systems they are responsible for… and they don’t even need coffee or time off.
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
By leveraging custom resources, Kubernetes principles such as the control loop and custom code, operators aim to provide the experience of a skilled maintainer of the service or software for which they are responsible. In essence, this allows not only the possibility of easy deployment of a given product or service, but also continued software health monitoring, upgrades, advanced deployment schemes, alerting, metrics, scaling, and more. 

Operators run alongside resources they manage in the same Kubernetes clusters, giving them direct control and monitoring of said resources. Think of them as being coded with the knowledge of a system admin that can make changes and manage systems they are responsible for… and they don’t even need coffee or time off.
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+Talking to other attendees, I found that what seemed like clear advantages of operator patterns vs. Helm charts often didn’t boil down to a substantial reason to choose operators in all cases. Much of this seemed to stem from a mixture of Helm being more well-known and understood, along with many operators not yet offering functionality that Helm didn’t already provide. I also noted that I approached the Helm vs operator conversation with something like management of a long-lived critical full-stack application in mind. When doing so, the advantages of an operator become much clearer. Knowing that your operator can handle something like a database schema migration check, or deploying your app with blue/green versioning, helps contrast the two options.
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
Talking to other attendees, I found that what seemed like clear advantages of operator patterns vs. Helm charts often didn’t boil down to a substantial reason to choose operators in all cases. Much of this seemed to stem from Helm being more well-known and understood, combined with many operators not yet offering functionality that Helm already provides. 

I also noted that I approached the Helm vs. operator conversation with something like management of a long-lived critical full-stack application in mind. When doing so, the advantages of an operator become much clearer. Knowing that your operator can handle something like a database schema migration check, or deploying your app with blue/green versioning, helps contrast the two options.
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+It goes without saying that we are just trying to follow an increasingly-common trend here. For this reason, I was happily surprised with the concepts presented by Goutham Veeramachaneni & Edward Welch in this talk:
+
+{{< youtube BmPnNmN9jtc >}}
+<br/>
+
+Their idea of a Monomicrolith really resonated. One of the struggles we have with the cloud server is that its responsibility to maintain our various cloud objects require it to be the source of truth when creating, updating, or tearing down infrastructure. This, in turn, means that a lot of configuration is defined through the server. The talk explored the idea that this sort of design isn’t necessarily a flaw. Instead, if the code is modular enough, you can specify what pieces of functionality each process should run on startup. This allows for most of the benefits of a single code-base and configuration style, while still allowing for extreme flexibility and horizontal scaling.
```
Comment (justinegeffen) [inline site/content/blog/2019-12-18-kubecon-na-2019.md]:
```suggestion
Their idea of a Monomicrolith really resonated. One of the struggles we have with the cloud server is that its responsibility to maintain our various cloud objects requires it to be the source of truth when creating, updating, or tearing down infrastructure. This, in turn, means that a lot of configuration is defined through the server. 

The talk explored the idea that this sort of design isn’t necessarily a flaw. Instead, if the code is modular enough, you can specify what pieces of functionality each process should run on startup. This allows for most of the benefits of a single code-base and configuration style, while still allowing for extreme flexibility and horizontal scaling.
```

File: site/content/blog/2019-12-18-kubecon-na-2019.md (hunk 7 above)
Comment (gabrieljackson) [reply site/content/blog/2019-12-18-kubecon-na-2019.md]: Sorry, I should maybe clarify, but the custom resource definitions have been around for a long time in alpha/beta. Kubernetes just likes to sit on its improvements for a long time before considering them "GA" level.

**Comment by bradjcoughlin** [comment]
Nice summary and great writing @gabrieljackson 👍

**Comment by gabrieljackson** [comment]
Thanks @bradjcoughlin!

**Pull request merged**